### PR TITLE
internal/extsvc/github: Testing improvements

### PR DESCRIPTION
**Pull request by indradhanush**
Stacked on top of #34172.

## Reviewer notes

This PR does the following:

1. Remove unused code and add debug log statements in [internal/extsvc/github/common.go](https://github.com/sourcegraph/sourcegraph/pull/34004/files#diff-4f26f74d24f2895f21a25955d52895028d99a1070aad8720638901dc691880f3).
1. Moves existing tests in [internal/extsvc/github/common_test.go](https://github.com/sourcegraph/sourcegraph/pull/34004/files#diff-f166190abda98b5ed2f24ec856b1bf687e4cad439aad812ab521def96f97a6cf) to their respective places in [internal/extsvc/github/v3_test.go](https://github.com/sourcegraph/sourcegraph/pull/34004/files#diff-17a4bd8d6b745f2a91fc022d2799f3c5227e69b0a61c7c406d4e7c7a712e17bc) and [internal/extsvc/github/v4_test.go](https://github.com/sourcegraph/sourcegraph/pull/34004/files#diff-5f1340d436e2a1b350de334673e9d5fbcf2f0feb43d20f4e6a068bd7512cf77b).
1. Adds new tests for `GetRepository` and improves testing for `ListOrganizations` for cached responses in [internal/extsvc/github/v3_test.go](https://github.com/sourcegraph/sourcegraph/pull/34004/files#diff-17a4bd8d6b745f2a91fc022d2799f3c5227e69b0a61c7c406d4e7c7a712e17bc).
1. Adds a helpful comment in [internal/extsvc/github/vcr_test.go](https://github.com/sourcegraph/sourcegraph/pull/34004/files#diff-5cfb00b8ed3042677f78352879c39450f5c6f6258db1d3b6aeee5bb3a5e7f79c).
1. Removes outdated comment and updates existing one in [internal/extsvc/github/v3.go](https://github.com/sourcegraph/sourcegraph/pull/34004/files#diff-b51b812bc803a0ab9412acfe7408347ba001f43e8f6a9bc211303a49a377a3fa).
1. Sets up a default HTTP transport cache in [internal/httpcli/client.go](https://github.com/sourcegraph/sourcegraph/pull/34004/files#diff-39024228d118dcefb9d91855eadc6faf8aa279ec71ef0f7c078981c37d3ecea8) and uses it in [internal/httptestutil/recorder.go](https://github.com/sourcegraph/sourcegraph/pull/34004/files#diff-d9c0b0a240ba20e03b4decc66703e52d7dec28eb32151853c702fc9f89c0d53a). 

## Test plan

No functional change. Updates and additions to testing and removal of outdated code and comments.

### CODE REVIEW

File: internal/extsvc/github/v3_test.go [hunk 1]
```diff
@@ -452,12 +452,35 @@ func TestGetOrganization(t *testing.T) {
...
+}
+
 // ListOrganizations is primarily used for GitHub Enterprise clients. As a result we test against
 // ghe.sgdev.org.  To update this test, access the GitHub Enterprise Admin Account (ghe.sgdev.org)
 // with username milton in 1password. The token used for this test is named sourcegraph-vcr-token
 // and is also saved in 1Password under this account.
 func TestListOrganizations(t *testing.T) {
-	t.Run("enterprise-integration-without-cache", func(t *testing.T) {
+	t.Run("enterprise-integration", func(t *testing.T) {
 		cli, save := newV3TestEnterpriseClient(t, "ListOrganizations")
```
Comment (bobheadxi) [inline internal/extsvc/github/v3_test.go:484]: If you look into this function it looks like the the caching middleware is no longer registered: https://github.com/sourcegraph/sourcegraph/blob/main/internal/httptestutil/recorder.go#L80

cc @davejrt

Comment (bobheadxi) [reply internal/extsvc/github/v3_test.go:484]: (though `httpcli` should probably export the preconfigured `CachedTransportOpt` for use in these cases)

Comment (bobheadxi) [reply internal/extsvc/github/v3_test.go:484]: The lack of that middleware would explain why the cache isn't being hit in tests (there is no cache). The cache is instantiated by default for prod users of `httpcli` however

Comment (bobheadxi) [reply internal/extsvc/github/v3_test.go:484]: looks fine with the cache set up in tests:

```
=== RUN   TestGetRepository
t=2022-04-19T15:02:43-0700 lvl=dbug msg="TRACE github" host=api.github.com path=/repos/sourcegraph/sourcegraph code=200 duration=483.520708ms
t=2022-04-19T15:02:43-0700 lvl=dbug msg=doRequest status="200 OK" x-ratelimit-remaining=4997
t=2022-04-19T15:02:43-0700 lvl=dbug msg="TRACE github" host=api.github.com path=/repos/sourcegraph/sourcegraph code=200 duration=3.931083ms
t=2022-04-19T15:02:43-0700 lvl=dbug msg=doRequest status="200 OK" x-ratelimit-remaining=4997
```

Also see the updated VCRs that demonstrate this in action, with `X-From-Cache`: https://github.com/sourcegraph/sourcegraph/pull/34004/commits/b3a0e430cc0526a975038c827769e9bd48a0971b

Comment (indradhanush) [reply internal/extsvc/github/v3_test.go:514]: Thank you for your work! This was a great find! <3

**Comment by indradhanush** [comment]
Outdated by #34161.

**Comment by bobheadxi** [comment]
@indradhanush well still want to merge these test updates at some point!

**Comment by indradhanush** [comment]
@bobheadxi Sorry my bad. I can update this PR today and get it ready for review.

File: internal/extsvc/github/common.go [hunk 2]
```diff
@@ -1543,6 +1538,8 @@ func doRequest(ctx context.Context, apiURL *url.URL, auth auth.Authenticator, ra
 	}
 	defer resp.Body.Close()
 
+	log15.Debug("doRequest", "status", resp.Status, "x-ratelimit-remaining", resp.Header.Get("x-ratelimit-remaining"))
```
Comment (bobheadxi) [inline internal/extsvc/github/common.go:1541]: Do we still need/want this?

**Review by bobheadxi — CHANGES_REQUESTED** [review]
@indradhanush I noticed you've re-generated your VCRs but none of the recordings have `X-From-Cache` anymore - did you have redis set up correctly? When I ran the tests previously I definitely saw `X-From-Cache` in the recordings

**Comment by indradhanush** [comment]
@bobheadxi Hmm. That's super weird. Even though the assertion for the rate limit verifies that the second request uses the cache. I had redis running locally if that's what you meant by having redis setup correctly. Is there an additional step that needs to be done? 

Note that I had to add the calls to rcache.SetupForTest in these tests to suppress the warnings related to redis connection not found.

**Comment by bobheadxi** [comment]
> Note that I had to add the calls to rcache.SetupForTest in these tests to suppress the warnings related to redis connection not found.

Ah if you've suppressed the warnings then maybe it's fine? I had to set an env variable to get it to use my cache

**Review by bobheadxi — APPROVED** [review]
If the test coverage is satisfactory let's go 😁 Thanks for your investigation!

File: internal/extsvc/github/common.go (hunk 2 above)
Comment (indradhanush) [reply internal/extsvc/github/common.go:1541]: I think it's useful, especially when running the tests. I'd weigh towards keeping it.

**Comment by indradhanush** [comment]
> Ah if you've suppressed the warnings then maybe it's fine? I had to set an env variable to get it to use my cache

Oh. What was the env? Should we be using that instead of `rcache.SetupForTests(t)`? 

Also if we're expecting the `X-From-Cache` header to be set and it isn't getting set currently, then isn't that unexpected behaviour?

**Comment by indradhanush** [comment]
@bobheadxi I'll go ahead and merge it and if we need to follow up with some investigations regarding the `X-From-Cache`, I'm happy to do that as a follow up PR.

**Pull request merged**